### PR TITLE
feat: Add support for multiple databases in Firestore

### PR DESCRIPTION
**Pull request by hemanshv**
add code to support configuration of multiple Db in firestore native mode

### CODE REVIEW

**Comment by jskeet** [comment]
Please could you change the first line of the commit to "feat: add support for multiple databases" or similar? That way it will be picked up in release notes. (Just run `git commit --amend` to amend the commit, then `git push -f` to push it.)

**Comment by hemanshv** [comment]
sure Jon

**Comment by hemanshv** [comment]
@jskeet  please review but do not merge, Expecting an integration test will fail because the feature is not launched yet.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+// limitations under the License.
+
+using Google.Cloud.ClientTesting;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Google.Cloud.Firestore.IntegrationTests
+{
+
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs:22]: Nit: please remove the blank line.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            FirestoreDb correctDbReader = new FirestoreDbBuilder { DatabaseId = correctDbId, ProjectId = _fixture.ProjectId }.Build();
+            DocumentReference readerDocRef = correctDbReader.Collection("cities").Document("SF");
+            DocumentSnapshot CorrectDbSnapshot = await readerDocRef.GetSnapshotAsync();
+            Dictionary<string, object> items = CorrectDbSnapshot.ToDictionary();
+            // Try to read data from random incorrect database. To validate that the snapshot and data should not exists.
+            FirestoreDb incorrectDbReader = new FirestoreDbBuilder { DatabaseId = incorrectDbId, ProjectId = _fixture.ProjectId }.Build();
+            DocumentReference incorrectDbDocRef = incorrectDbReader.Collection("cities").Document("SF");
+            DocumentSnapshot incorrectDbSnapshot = await incorrectDbDocRef.GetSnapshotAsync();
+
+            Assert.Equal(true, CorrectDbSnapshot.Exists);
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs:61]: I'd also suggest moving the assertions to just after you've done the conversion.
(In some ways it would be better for this to be two tests, but I can see why it's only one.)

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/Google.Cloud.Firestore.IntegrationTests.csproj [hunk 3]
```diff
@@ -7,8 +7,9 @@
     <NoWarn>1701;1702;1705;xUnit2004;xUnit2013</NoWarn>
   </PropertyGroup>
   <ItemGroup>
-    <PackageReference Include="Google.Api.Gax.Grpc.Testing" Version="[4.0.0, 5.0.0)" />
-    <PackageReference Include="Google.Api.Gax.Testing" Version="[4.0.0, 5.0.0)" />
+    <PackageReference Include="Google.Api.Gax.Grpc.Testing" Version="[3.7.0, 4.0.0)" />
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/Google.Cloud.Firestore.IntegrationTests.csproj:10]: These are out of date - please pull the upstream main branch and rebase on top of it, removing unnecessary changes.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        // Credentials scope, To view and manage your Google Cloud Datastore data.
+        private const string FirestoreCredentialScope = "https://www.googleapis.com/auth/datastore";
+        private const string ProjectEnvironmentVariable = "FIRESTORE_TEST_PROJECT";
+        private static readonly string s_credentialsPath = Environment.GetEnvironmentVariable("GOOGLE_APPLICATION_CREDENTIALS");
+        private static GoogleCredential s_googleCredential;
+        private static HttpClient s_httpClient;
+
+        static MulitpleFirestoreDbFixture()
+        {
+            // Initalize gcloud credentials to be used while REST Api calls.
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs:40]: "while REST Api" => "with REST API"

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                });
+            }
+            s_httpClient = new HttpClientFactory()
+                .CreateHttpClient(new CreateHttpClientArgs()
+               {
+                   ApplicationName = "Firestore integration tests",
+                   GZipEnabled = true,
+                   Initializers = { s_googleCredential },
+               });
+
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs:56]: Nit: remove blank line.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        {
+        }
+
+        // creates new firestore database currently using rest API once the Admin Apis are available will be migrated to Grpc.
+        public async Task CreateDatabaseAsync(string databaseId, string locationId, string databaseType)
+        {
+            var createDatabaseEndpoint = new Uri($"{FirestoreURl}/projects/{ProjectId}/databases?database_id={databaseId}");
+            var data = new { locationId = locationId, type = databaseType };
+            var httpData = new StringContent(JsonConvert.SerializeObject(data), Encoding.UTF8, "application/json");
+            await s_httpClient.PostAsync(createDatabaseEndpoint, httpData);
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs:69]: You should probably check that the response is successful.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+                   Initializers = { s_googleCredential },
+               });
+
+        }
+
+        public MulitpleFirestoreDbFixture() : base(ProjectEnvironmentVariable)
+        {
+        }
+
+        // creates new firestore database currently using rest API once the Admin Apis are available will be migrated to Grpc.
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs:63]: This comment is a little unclear at the moment (and it starts with a lower case letter - tiny nit).

I would suggest making this a regular XML comment:

```csharp
// TODO: Use Google.Cloud.Firestore.Admin.V1 when that supports this operation.

/// <summary>
/// Creates a new Firestore database using the REST API.
/// </summary>
```

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FirestoreDbBuilderTest.cs [hunk 8]
```diff
@@ -45,6 +45,27 @@ public void MaximalConfiguration()
             Assert.IsType<CustomConverter<Guid>>(db.SerializationContext.GetConverter(typeof(Guid)));
         }
 
+        [Fact]
+        public void NamedDatabase()
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FirestoreDbBuilderTest.cs:49]: This is a unit test, but it's using the application default credentials. It's probably simplest to set the CallInvoker to a FakeCallInvoker in the FirestoreDbBuilder below.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FirestoreDbBuilderTest.cs [hunk 9]
```diff
@@ -45,6 +45,27 @@ public void MaximalConfiguration()
...
+                DatabaseId = databaseId
+            }.Build();
+
+            FirestoreDb db = new FirestoreDbBuilder
+            {
+                ProjectId = projectId,
+                DatabaseId = null
+            }.Build();
+
+            Assert.Equal(databaseId,namedDb.DatabaseId);
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FirestoreDbBuilderTest.cs:65]: I'd split this one test into two: DefaultDatabase (and don't set the DatabaseId explicitly), and NonDefaultDatabase.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestoreDb.cs [hunk 10]
```diff
@@ -43,11 +43,10 @@ public sealed class FirestoreDb
         /// </summary>
         public string ProjectId { get; }
 
-        // Note: currently internal as only the default database ID is supported.
         /// <summary>
         /// The database ID associated with this database.
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestoreDb.cs:47]: Add comment saying that the value will be "(default)" for the project-default database?

**Review by jskeet — COMMENTED** [review]
Mostly fine - lots of comment nits, but they should be easy to fix.

**Comment by hemanshv** [comment]
Thanks Jon for the review will address the review comments.

**Comment by jskeet** [comment]
> Merge branch 'googleapis:main' into firestoreMulitpleDB

Rather than merging main into the feature branch, it's better to pull main, then rebase the feature branch onto it. We can go over this in the retrospective later.

**Comment by hemanshv** [comment]
yes Jon Sure

**Comment by hemanshv** [comment]
@jskeet  I have addressed the review comments, Please review.
Thanks,

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FirestoreDbBuilderTest.cs [hunk 11]
```diff
@@ -1,66 +1,93 @@
...
-
-            public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request) =>
-                throw new NotImplementedException();
-
-            public override TResponse BlockingUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request) =>
-                throw new NotImplementedException();
-        }
-    }
-}
+// Copyright 2019, Google LLC
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FirestoreDbBuilderTest.cs:1]: I'm not sure why the whole of this file is showing up as changed... did you edit this in a different way to the other files? I assume it's a line-ending change somehow, but I don't know what would cause it. It might be worth copying the new tests, reverting this file, then editing it again.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FirestoreDbBuilderTest.cs [hunk 12]
```diff
@@ -1,66 +1,93 @@
...
+
+        [Fact]
+        public void DefaultDatabase()
+        {
+            string projectId = "testProjectId";
+            FirestoreDb defaultDb = new FirestoreDbBuilder
+            {
+                ProjectId = projectId,
+                CallInvoker = new FakeCallInvoker()
+
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FirestoreDbBuilderTest.cs:56]: Please remove this blank line.

**Review by jskeet — APPROVED** [review]
Basically fine, just a few formatting nits.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Cloud.ClientTesting;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Google.Cloud.Firestore.IntegrationTests
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs:20]: I should have noticed this before - as this is a new file, let's use file-scoped namespaces. Just add ; at the end of this line, and VS will do the rest. It'll reduce indentation. Ditto the fixture.

**Comment by hemanshv** [comment]
@jskeet fix the nits please review 
Thanks,

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FirestoreDbBuilderTest.cs [hunk 14]
```diff
@@ -1,66 +1,90 @@
...
+        };
+        var db = builder.Build();
+        Assert.Equal("proj", db.ProjectId);
+        Assert.Equal("db", db.DatabaseId);
+        db.LogWarning("Test warning");
+        Assert.Equal("Test warning", warning);
+        Assert.IsType<FakeClock>(db.Client.Settings.Clock);
+        Assert.IsType<CustomConverter<Guid>>(db.SerializationContext.GetConverter(typeof(Guid)));
+    }
+    [Fact]
```
[the remark this answers is not in the document]
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/FirestoreDbBuilderTest.cs:48]: Hey @jskeet ,
Thanks for reviewing the PR. I have fixed the nit.
Will the raise the PR next, as the feature is not GA yet, we will not merge it.
Thanks,

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+using System;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Google.Cloud.Firestore.IntegrationTests;
+
+[CollectionDefinition(nameof(MulitpleFirestoreDbFixture))]
+public class MulitpleFirestoreDbFixture : CloudProjectFixtureBase, ICollectionFixture<MulitpleFirestoreDbFixture>
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs:28]: Please rename to MultipleFirestoreDbFixture (and rename the file)

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    
+    [Fact]
+    public async Task MultipleDbReadWriteTest()
+    {
+        string dbIdPrefix = IdGenerator.FromGuid(prefix: "db", separator: "-");
+        string correctDbId = $"{dbIdPrefix}-correct";
+        string incorrectDbId = $"{dbIdPrefix}-incorrect";
+        string databaseLocationId = "us-east1";
+        string databaseType = "FIRESTORE_NATIVE";
+        // Create two databases one for reading and writing, one for reading to validate that correct database should be referenced.
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs:38]: I'd suggest adding a blank line above each of the comments, just to split the test up a bit.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+        await _fixture.CreateDatabaseAsync(correctDbId, databaseLocationId, databaseType);
+        await _fixture.CreateDatabaseAsync(incorrectDbId, databaseLocationId, databaseType);
+        // Write data to the first database; referenced as correct database.
+        FirestoreDb correctDbWriter = new FirestoreDbBuilder { DatabaseId = correctDbId, ProjectId = _fixture.ProjectId }.Build();
+        await correctDbWriter.Collection("cities").Document("SF").SetAsync(new Dictionary<string, object>{
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs:43]: Please move the `{` onto the start of the next line (on a new line on its own) just for balance. (It may be like this elsewhere, of course...)

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+        });
+        // Read data from database that was referenced while writing.
+        FirestoreDb correctDbReader = new FirestoreDbBuilder { DatabaseId = correctDbId, ProjectId = _fixture.ProjectId }.Build();
+        DocumentReference readerDocRef = correctDbReader.Collection("cities").Document("SF");
+        DocumentSnapshot correctDbSnapshot = await readerDocRef.GetSnapshotAsync();
+        Dictionary<string, object> items = correctDbSnapshot.ToDictionary();
+        Assert.True(correctDbSnapshot.Exists);
+        Assert.Equal(5, items.Count);
+
+        // Try to read data from random incorrect database, to validate that the snapshot and data does not exists.
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs:58]: "does not exists" => "do not exist"

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+{
+    private const string RestApiUrlBase = "https://firestore.googleapis.com/v1";
+    // Credentials scope, To view and manage your Google Cloud Datastore data.
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs:31]: I'd probably delete the comment - or maybe rewrite it as "Scope used for the REST API to create databases"

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        // Initalize gcloud credentials to be used with REST API calls. 
+        s_googleCredential = GoogleCredential.FromFile(s_credentialsPath);
+        if (s_googleCredential.IsCreateScopedRequired)
+        {
+            s_googleCredential = s_googleCredential.CreateScoped(new[]
+            {
+                FirestoreCredentialScope
+            });
+        }
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs:48]:
```suggestion
        s_googleCredential = GoogleCredential.FromFile(s_credentialsPath).CreateScoped(FirestoreCredentialScope);
```

(There's no need to conditionalize it, and it uses a params array.)

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+        s_httpClient = new HttpClientFactory()
+            .CreateHttpClient(new CreateHttpClientArgs()
+           {
+               ApplicationName = "Firestore integration tests",
+               GZipEnabled = true,
+               Initializers = { s_googleCredential },
+           });
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs:55]:
```suggestion
        s_httpClient = new HttpClientFactory()
            .CreateHttpClient(new CreateHttpClientArgs { Initializers = { s_googleCredential } });
```

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+
+[CollectionDefinition(nameof(MulitpleFirestoreDbFixture))]
+public class MulitpleFirestoreDbFixture : CloudProjectFixtureBase, ICollectionFixture<MulitpleFirestoreDbFixture>
+{
+    private const string RestApiUrlBase = "https://firestore.googleapis.com/v1";
+    // Credentials scope, To view and manage your Google Cloud Datastore data.
+    private const string FirestoreCredentialScope = "https://www.googleapis.com/auth/datastore";
+    private const string ProjectEnvironmentVariable = "FIRESTORE_TEST_PROJECT";
+    private static readonly string s_credentialsPath = Environment.GetEnvironmentVariable("GOOGLE_APPLICATION_CREDENTIALS");
+    private static readonly GoogleCredential s_googleCredential;
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs:35]: Do we actually need this? It's built into the HttpClient anyway, so I don't think we need it as a field. Just use a local variable in the static constructor.

**Review by jskeet — COMMENTED** [review]
Mostly fine, just a few nits.

Please rebase your branch to main as part of the update though - it's currently 300+ commits behind.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs (hunk 22 above)
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs:35]: Apart from ProjectEnvironmetVariable others can be moved to the constructor. As ProjectEnvironmentVariable is passed to the base class constructor.

Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MulitpleDatabaseFirestoreFixture.cs:35]: It's fine for the constants to stay where they are - but yes, you could move the credential path to be a local variable too.

**Comment by hemanshv** [comment]
@jskeet I have addressed the review comments. Please have a look and provide your feedback.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+        await _fixture.CreateDatabaseAsync(correctDbId, databaseLocationId, databaseType);
+        await _fixture.CreateDatabaseAsync(incorrectDbId, databaseLocationId, databaseType);
+
+        // Write data to the first database; referenced as correct database.
+        FirestoreDb correctDbWriter = new FirestoreDbBuilder
+        {
+            DatabaseId = correctDbId,
+            ProjectId = _fixture.ProjectId
+        }.Build();
+        await correctDbWriter.Collection("cities").Document("SF").SetAsync(new Dictionary<string, object>{
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs:49]: I'd still like to see this open brace on a new line. (Or at the very least, a space before it - but it would look better on a new line IMO.)

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MultipleFirestoreDbFixture.cs [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+    static MultipleFirestoreDbFixture()
+    {
+        // Scope used for the REST API to create databases.
+        string FirestoreCredentialScope = "https://www.googleapis.com/auth/datastore";
+        string s_credentialsPath = Environment.GetEnvironmentVariable("GOOGLE_APPLICATION_CREDENTIALS");
+
+        // Initalize gcloud credentials to be used with REST API calls. 
+        GoogleCredential s_googleCredential = GoogleCredential.FromFile(s_credentialsPath).CreateScoped(FirestoreCredentialScope);
+        s_httpClient = new HttpClientFactory()
+            .CreateHttpClient(new CreateHttpClientArgs() { Initializers = { s_googleCredential } });
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MultipleFirestoreDbFixture.cs:43]: Nit: remove the `()` from `new CreateHttpClientArgs()`

**Review by jskeet — APPROVED** [review]
Just a couple of formatting nits left.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs (hunk 23 above)
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs]: Done

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MultipleFirestoreDbFixture.cs (hunk 24 above)
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MultipleFirestoreDbFixture.cs]: Done

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs [hunk 25]
```diff
@@ -0,0 +1,76 @@
...
+using Google.Cloud.ClientTesting;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Google.Cloud.Firestore.IntegrationTests;
+
+public class FirestoreMultipleDbTest
+{
+    private readonly MultipleFirestoreDbFixture _fixture;
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs:24]: Sorry, I hadn't spotted this before.

This isn't how fixtures are meant to be used. The point of a fixture is for Xunit to instantiate it and pass it to the test. That way the same object can be used for multiple tests etc.

The test class should have the `[Collection(nameof(MultipleFirestoreDbFixture))]` attribute, and it should accept the fixture in a constructor. See RunInTransactionTest for an example.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MultipleFirestoreDbFixture.cs [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+public class MultipleFirestoreDbFixture : CloudProjectFixtureBase, ICollectionFixture<MultipleFirestoreDbFixture>
+{
+    private const string RestApiUrlBase = "https://firestore.googleapis.com/v1";
+    private const string ProjectEnvironmentVariable = "FIRESTORE_TEST_PROJECT";
+    private static readonly HttpClient s_httpClient;
+
+    static MultipleFirestoreDbFixture()
+    {
+        // Scope used for the REST API to create databases.
+        string FirestoreCredentialScope = "https://www.googleapis.com/auth/datastore";
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MultipleFirestoreDbFixture.cs:37]: I'd rename this to just `scope`, given the closer context.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MultipleFirestoreDbFixture.cs [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+        string s_credentialsPath = Environment.GetEnvironmentVariable("GOOGLE_APPLICATION_CREDENTIALS");
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MultipleFirestoreDbFixture.cs:38]: Please rename to `credentialsPath` - local variables should never have a `_` or `s_` prefix.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MultipleFirestoreDbFixture.cs [hunk 28]
```diff
@@ -0,0 +1,63 @@
...
+
+        // Initalize gcloud credentials to be used with REST API calls. 
+        GoogleCredential s_googleCredential = GoogleCredential.FromFile(s_credentialsPath).CreateScoped(FirestoreCredentialScope);
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MultipleFirestoreDbFixture.cs:41]: Ditto

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MultipleFirestoreDbFixture.cs [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+using Xunit;
+
+namespace Google.Cloud.Firestore.IntegrationTests;
+
+[CollectionDefinition(nameof(MultipleFirestoreDbFixture))]
+public class MultipleFirestoreDbFixture : CloudProjectFixtureBase, ICollectionFixture<MultipleFirestoreDbFixture>
+{
+    private const string RestApiUrlBase = "https://firestore.googleapis.com/v1";
+    private const string ProjectEnvironmentVariable = "FIRESTORE_TEST_PROJECT";
+    private static readonly HttpClient s_httpClient;
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MultipleFirestoreDbFixture.cs:32]: Given that we only ever need to use this from an instance context, I wonder whether it would make sense to have this as an instance field, and move all the initialization into the constructor. That's *usually* nicer than having a static constructor. It doesn't matter too much though.

**Review by jskeet — COMMENTED** [review]
Sorry, there are fixture aspects I hadn't noticed before. Nearly there though.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MultipleFirestoreDbFixture.cs (hunk 29 above)
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MultipleFirestoreDbFixture.cs:32]: These are made static to basically share the HTTP client object between multiple instances. Although its a fixture that will be shared by multiple tests. I think it make sense to convert them to instance variables. 
Thanks.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs (hunk 25 above)
Comment (hemanshv) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/FirestoreMultipleDbTest.cs:24]: Thanks @jskeet for pointing this out. Will update it.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MultipleFirestoreDbFixture.cs [hunk 30]
```diff
@@ -27,24 +27,22 @@ namespace Google.Cloud.Firestore.IntegrationTests;
 [CollectionDefinition(nameof(MultipleFirestoreDbFixture))]
 public class MultipleFirestoreDbFixture : CloudProjectFixtureBase, ICollectionFixture<MultipleFirestoreDbFixture>
 {
-    private const string RestApiUrlBase = "https://firestore.googleapis.com/v1";
+    private readonly string _restApiUrlBase;
```
Comment (jskeet) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.IntegrationTests/MultipleFirestoreDbFixture.cs:30]: It looks like this really is a constant, so can probably stay as it was.

**Review by jskeet — APPROVED** [review]
I'd make the REST URI a constant again, but other than that, this looks good to go.

**Review by jskeet — APPROVED** [review]
Much better, thanks. I'd suggest squashing it all down to one commit at this point.
(And IIRC, we can't actually merge yet?)

**Comment by hemanshv** [comment]
Right @jskeet we cannot merge it as the feature is'nt GA yet.

**Comment by jskeet** [comment]
Importantly, the diff now shows that there are no API changes :)
